### PR TITLE
Allow setting single key shortcuts

### LINKED ISSUE #29: Allow setting single key shortcuts [$20]
According to what is said at https://bugs.launchpad.net/switchboard-plug-keyboard/+bug/1357895/comments/52, the modifier only shortcuts can be set in two different ways:

1. Adding both "<Modifier1>Modifier2_L" and "<Modifier1>Modifier2_R" to shortcut we are assigning. (Instead of "<Modifier1><Modifier2>". BTW, other desktop environments seem to use similar solution.)
2. Add one of the above based on the held keys when assigning shortcuts.

I think the first solution is more preferable because users don't consider any differences between Left or Right modifier keys and expect shortcuts to work using either of them.

Launchpad Details: [#LP1553555](https://bugs.launchpad.net/bugs/1553555) Ash - 2016-03-05 18:31:11 +0000

<bountysource-plugin>

---
There is a **[$20 open bounty](https://www.bountysource.com/issues/43871241-allow-setting-single-key-shortcuts?utm_campaign=plugin&utm_content=tracker%2F59531540&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F59531540&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by geocool**
Fixes #29

Allows binding single keys that won't interfere with typing.
(F1-F12, Media Keys etc)

### CODE REVIEW

File: src/Shortcuts/Shortcut.vala [hunk 1]
```diff
@@ -103,9 +103,44 @@ namespace Pantheon.Keyboard.Shortcuts
         // validator
         public bool valid()
         {
-            if (accel_key == 0 || (modifiers == (Gdk.ModifierType) 0 && accel_key != Gdk.Key.Print))
+            if (accel_key == 0)
                 return false;
 
+            if (modifiers == (Gdk.ModifierType) 0) 
+            {
```
Comment (jeremypw) [inline src/Shortcuts/Shortcut.vala]: Elementary code style - brace should be on same line as if clause.

File: src/Shortcuts/Shortcut.vala [hunk 2]
```diff
@@ -103,9 +103,44 @@ namespace Pantheon.Keyboard.Shortcuts
...
+                || (accel_key == Gdk.Key.bracketright)
+                || (accel_key == Gdk.Key.bracketleft)
+                || (accel_key == Gdk.Key.apostrophe)
+                || (accel_key == Gdk.Key.semicolon)
+                || (accel_key == Gdk.Key.slash)
+                || (accel_key == Gdk.Key.period)
+                || (accel_key == Gdk.Key.comma)
+                || (accel_key == Gdk.Key.grave)
+                )
+               {
```
Comment (jeremypw) [inline src/Shortcuts/Shortcut.vala]: Elementary code style - brace should be on same line as closing bracket.

File: src/Shortcuts/Shortcut.vala [hunk 3]
```diff
@@ -103,9 +103,44 @@ namespace Pantheon.Keyboard.Shortcuts
         // validator
         public bool valid()
         {
-            if (accel_key == 0 || (modifiers == (Gdk.ModifierType) 0 && accel_key != Gdk.Key.Print))
+            if (accel_key == 0)
                 return false;
 
+            if (modifiers == (Gdk.ModifierType) 0)
```
Comment (jeremypw) [inline src/Shortcuts/Shortcut.vala]: Shouldn't the SHIFT_MASK be retained here?

**Comment by jeremypw** [comment]
You could save some lines by combining the common parts of the two if clauses.

**Comment by jeremypw** [comment]
I wonder whether it is possible to use unichar.isprint () function to exclude invalid characters?

**Comment by geocool** [comment]
Fixed my code to follow elementary code style and I reduced the code as you suggested.
My previous commit followed the file's code style which must be outdated. 

I don't know if an implementation with isprint is possible, I could try it next weekend and do another pull request refactoring valid function.

**Comment by jeremypw** [comment]
Thanks.  You are right that much of the existing code pre-dates current elementary code style.  At some point a PR fixing that will have to be produced; new code should comply.

**Comment by jeremypw** [comment]
Not sure why this is not passing Travis CI but I cannot merge until it does.

**Comment by FernandoGarcia** [comment]
Thank you very much!!

Elementary merge it please.

**Comment by geocool** [comment]
I read somewhere that forks need to enable travis CI too.
I enabled Travis CI in my fork and triggered a build. 
I don't know if there is anything else to do in order to fix approval on my part .
If there is, let me know.

**Comment by jeremypw** [comment]
@geocool Make sure you use `https://travis-ci.com` not `https://travis-ci.org`.  I'll check the elementary admins whether they need to change any settings on this repo.

**Pull request merged**